packages/core/http/core-http-server-internal/src/base_path_service.ts

```diff
@@ -28,19 +36,47 @@ export class BasePath implements IBasePath {
   }
 
   public get = (request: KibanaRequest) => {
-    const requestScopePath = this.basePathCache.get(ensureRawRequest(request)) || '';
+    const cacheValue = this.basePathCache.get(ensureRawRequest(request));
+    const requestScopePath = isObject(cacheValue)
+      ? Object.values(cacheValue)
+          .sort((a, b) => a.index - b.index)
+          .map((v) => v.basePath)
+          .join('')
+      : cacheValue || '';
```

Comment:
I'm sorry, I did read the PR's description, so I get what you're trying to do, but I'm still going to have to ask you to explain exactly why all of this was strictly necessary 🙂 

- What are we accomplishing by having the solutionId in the base path?
- Why can't we have a single source of truth to set this basePath in our basePath map?

Reply:
> What are we accomplishing by having the solutionId in the base path?

Having the solutionId in the base path allows us to know which solution the user was when refreshing the browser window. The "discover" app is in 3 solutions, when refreshing the page we want to render that same solution the user was.
It also serve when sharing a link with other users. The other users will open that app in the same solution context.

> Why can't we have a single source of truth to set this basePath in our basePath map?

I followed what was in place with spaces and allowed multiple parts of Kibana to define the base path for a specific Request. We are kind of having the same problem here and spaces, so building up on top of what is already implemented seemed logical.
How different would you do it?

Reply:
Thanks for the explanations 

The thing is, the basePath is not only used for base for the "user base location in the app", it's also used by the browser to use as base for all API requests performed against the server. 

For spaces this was absolutely something we wanted, given we needed the ajax requests to be performed against that same "basePath" to preserve the information of the space they were performed from, but I'm not sure this is the case here, is it?

To rephrase, by changing our browser's base path from, say, `/[space]` to `/[space]/[solution]`, we're also going to have all ajax requests from the browser be performed against `[server]/[space]/[solution]/[api]` (instead of `[server]/[space]/[api]`), right?

First thing coming to mind is that the base path was supposed to be "immutable" during a browser's navigation. For spaces, switching from one space to the other can only be done via a full page refresh, causing the "immutable" basePath to change when a user navigate from a space to another. 

I assume this is not the same when navigation between solutions (within a given space), is it? You can enter Kibana on `solutionA`, then navigate to `solutionB` without triggering a page reload, right? Because in that case, the `basePath` on the browser-side won't change, so the requests will still fire against `/[space]/solutionA`, which seems arguably wrong?

Overall, I'm very fine trying to find a solution to add more context bound to the current navigation if we need to do so, I just wonder if the basePath really is the right place to do this. 

Also curious to hear @azasypkin feeling on that one.

Reply:
> I assume this is not the same when navigation between solutions (within a given space), is it? You can enter Kibana on solutionA, then navigate to solutionB without triggering a page reload, right? Because in that case, the basePath on the browser-side won't change, so the requests will still fire against /[space]/solutionA, which seems arguably wrong?

Wrong assumption 😊 I followed the exact same pattern as with space selector (be it the page selector page or the dropdown menu). There **is a full page refresh** whenever the user changes solution. For the exact reason you are pointing out. It's the desired behaviour to later be able to apply telemetry to know under which solution certain pages (maybe APIs) were accessed.

So to your point: the base path is immutable during a browser session with the current implementation.

Reply:
cc @kevinsweet in case you want to add more context

Reply:
> Curious to hear @azasypkin and @rudolf's feelings on that one.

Just to get the full picture before suggesting anything:

- Do we have a well-defined definition of what a "solution" is?
- Does this change apply across all offerings (on-prem, ESS/ECE, Serverless), or is it specific to Serverless?
- Is the "solution" context necessary on both client and server sides, or is it only required on the client side?

Apart from these questions, I'm curious about the alternative options we've explored before settling on this solution and why they weren't viable. Apologies if this has been covered elsewhere, but I might have missed it. I'm asking because the base path service is a platform-level component that could be challenging to handle properly. Considering the potential risks and long-term maintenance, it seems reasonable to extend it only if: a) the change impacts all or most of our deployment models (offerings), b) it's required for both client and server-side contexts, and c) other options were worse in terms of risk or complexity. Spaces ticked all these boxes back then, let's make sure this change ticks them as well.

Reply:
> Do we have a well-defined definition of what a "solution" is?

Yes. In the upcoming stateful offering, the navigation system will be divided in 3 solutions: "search", "observability" and "security".

> Does this change apply across all offerings (on-prem, ESS/ECE, Serverless), or is it specific to Serverless?

No this change does not apply to the Serverless offering (as serverless already has the concept of "projects" which are one of the 3 solutions mentioned in previous point. So this change applies accross on-prem and ESS.

> Is the "solution" context necessary on both client and server sides, or is it only required on the client side?

It is necessary both on the client and server side. On the client we read it to know which navigation tree to load, on the server to redirect "non solution aware path" (e.g. /app/discover) to solution aware path (/n/security/app/discover). As changing the base path requires a full page reload we want to eagerly set one of the solution to avoid a double page refresh (1 on page load to set a default solution and 2 if we detect that the page does not belong to the correct solution and we do a redirect to the correct solution. e.g. /n/es/app/observabilityOnboarding does not belong to "es". We detect it and redirect to "/n/oblt/app/observabilityOnboarding").

> Apart from these questions, I'm curious about the alternative options we've explored before settling on this solution and why they weren't viable.

Initially we explored saving the the latest solution a user was in in their user profile but it did not allow sharing a URL and maintain the context in which the app should be opened (e.g. the "discover" app can be in any of the 3 solutions).

Reply:
@sebelga How many apps/paths are solution aware/have a solution-specific render?

Reply:
@pgayvallet @rudolf & @azasypkin - Can one of you gents contextualize for me what would break by adding this "solution" section to the base path?  I understand the telemetry implications, but you lost me with the Ajax API.  

Whatever breaks, it needs to be addressed, but here is some more info - The first "release" of this new left nav will be locked behind Launch Darkly and be available only to Elastician cloud instances on the QA environment.  We are doing this to identify exactly these kinds of issues, and to allow many more internal folks to dogfood this and give feedback before it hits production.  Knowing that only the "old" UX will be accessible on production deployments and staging, do we need to solve the base URL path problem now, or can we identify it as something that must be fixed prior to a real launch, but allow it in the "internal preview" on QA?

Reply:
@petrklapka @sebelga I think we need to explain in a document what requirements we are solving and why this is the solution we feel fits the best. I don't think we can have this conversation in a GH issue effectively and we need a writeup to clearly articulate the problem, possible solutions we considered, and the pros and cons.

Reply:
> How many apps/paths are solution aware/have a solution-specific render?

@jeramysoucy  Let me restate the problem:

We need a way, **in the URL** to know if "/app/discover" should open inside the "search", "observability" or "security" solution. And we need it for _any_ plugin accros Kibana.

I only know 2 ways of doing this:

* prepend the path (base path approach): `/n/security/app/discover`
* query param: `/app/discover?solution=security`

Going the query param way would mean to go everywhere in Kibana where URL are constructed and make sure we forward the solution... I really don't think we want to go that route. Aside, we already solved this problem for spaces in a specific way (basepath).

@rayafratkina before going the doc route, let's make sure we all have the same requirements in mind.

These are the requirements I received:

- preserve the solution when refreshing the page
- being able to share a URL that opens a plugin inside a specific solution (so 2 users see the exact same page)

If those are correct, the solution needs to be persisted in the URL _somewhere_.

BTW, teams are already noticing the problem (https://github.com/elastic/kibana/pull/181079#issuecomment-2074621084).